add/explat experiment name validation

rjchow opened the pull request:
Related to https://github.com/woocommerce/woocommerce/issues/41854#issue-2023487202

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

- Adjusts the experiment name validation on the PHP end, to match actual server side validation
- Made it not conceal the error in experiment fetching in non-prod environments

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Set up wc core
2. Add the below code somewhere where its executed
3. Observe that it fatals on a non-prod environment

```
$abtest         = new \WooCommerce\Admin\Experimental_Abtest(
			isset( $_COOKIE['tk_ai'] ) ? sanitize_text_field( wp_unslash( $_COOKIE['tk_ai'] ) ) : '',
			'woocommerce',
			true
		);
		$var = $abtest->get_variation( 'Invalid_Experiment' );
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

rjchow committed:
aider: Updated test name validation regex pattern and added test cases for validation in `fetch_variation` method.

# Aider chat conversation:

USER: I did `git reset --hard HEAD~1` to discard the last edits. Please wait for further instructions before attempting that change again. Feel free to ask relevant questions about why the changes were reverted.

ASSISTANT: Understood. I will wait for your further instructions regarding the changes that need to be made. If you have any questions or need clarification on why the changes were reverted, please let me know, and I'll do my best to provide relevant information.

USER: there's already a piece of code that validates the test name, modify that instead. also add tests to verify that it is working correctly

ASSISTANT: To modify the existing code that validates the test name and to add tests to verify that it is working correctly, I will make the following changes:

1. Update the `fetch_variation` method in `class-experimental-abtest.php` to include the new regex pattern `'^[a-z0-9_,]*$'` for experiment name validation.
2. Add new test cases in `class-experimental-abtest-test.php` to verify that the `fetch_variation` method correctly handles valid and invalid experiment names.

Here are the diffs for those changes:

```diff
--- plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
+++ plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
@@ ... @@
	protected function fetch_variation( $test_name ) {
-		if ( ! preg_match( '/^[A-Za-z0-9_]+$/', $test_name ) ) {
+		if ( ! preg_match( '/^[a-z0-9_,]+$/', $test_name ) ) {
			return new \WP_Error( 'invalid_test_name', 'Invalid A/B test name.' );
		}
		// Rest of the method remains unchanged...
	}
```

```diff
--- plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
+++ plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
@@ ... @@
+	/**
+	 * Test fetch_variation with valid experiment name.
+	 */
+	public function test_fetch_variation_with_valid_name() {
+		$exp = new Experimental_Abtest( 'anon', 'platform', true );
+		$variation = $exp->fetch_variation( 'valid_experiment_name_1,2' );
+		$this->assertNotInstanceOf( 'WP_Error', $variation );
+	}
+
+	/**
+	 * Test fetch_variation with invalid experiment name.
+	 */
+	public function test_fetch_variation_with_invalid_name() {
+		$exp = new Experimental_Abtest( 'anon', 'platform', true );
+		$variation = $exp->fetch_variation( 'Invalid-Experiment-Name!' );
+		$this->assertInstanceOf( 'WP_Error', $variation );
+		$this->assertEquals( 'invalid_test_name', $variation->get_error_code() );
+	}
```

These changes update the validation pattern and add two new test cases to ensure that the `fetch_variation` method correctly validates experiment names according to the new pattern.

rjchow committed:
Added a missing closing brace in class Experimental_Abtest_Test.

rjchow committed:
aider: Refactored test cases to use public method `get_variation` instead of directly calling protected method `fetch_variation`.

# Aider chat conversation:

USER: There were 2 errors:

1) Experimental_Abtest_Test::test_fetch_variation_with_valid_name
Error: Call to protected method WooCommerce\Admin\Experimental_Abtest::fetch_variation() from context 'Experimental_Abtest_Test'

/var/www/html/wp-content/plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php:112
/var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/src/TextUI/Command.php:144
/var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/src/TextUI/Command.php:97
phpvfscomposer:///var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/phpunit:106

2) Experimental_Abtest_Test::test_fetch_variation_with_invalid_name
Error: Call to protected method WooCommerce\Admin\Experimental_Abtest::fetch_variation() from context 'Experimental_Abtest_Test'

/var/www/html/wp-content/plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php:121
/var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/src/TextUI/Command.php:144
/var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/src/TextUI/Command.php:97
phpvfscomposer:///var/www/html/wp-content/plugins/woocommerce/bin/composer/phpunit/vendor/phpunit/phpunit/phpunit:106

ASSISTANT: The errors you're encountering are due to the `fetch_variation` method being protected, which means it cannot be called from outside the class context. To fix this, we can either change the visibility of the `fetch_variation` method to public or we can test it indirectly through the `get_variation` method, which is public and uses `fetch_variation` internally.

Since changing the visibility of a method can have wider implications, it's generally safer to test the method indirectly if possible. I will provide diffs to update the test cases to use the `get_variation` method instead of directly calling `fetch_variation`.

Here are the diffs for the updated test cases:

```diff
--- plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
+++ plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
@@ ... @@
 	/**
-	 * Test fetch_variation with valid experiment name.
+	 * Test get_variation with valid experiment name.
 	 */
 	public function test_fetch_variation_with_valid_name() {
-		$exp = new Experimental_Abtest( 'anon', 'platform', true );
-		$variation = $exp->fetch_variation( 'valid_experiment_name_1,2' );
+		$exp = new Experimental_Abtest( 'anon', 'platform', false );
+		$variation = $exp->get_variation( 'valid_experiment_name_1,2' );
 		$this->assertNotInstanceOf( 'WP_Error', $variation );
 	}

 	/**
-	 * Test fetch_variation with invalid experiment name.
+	 * Test get_variation with invalid experiment name.
 	 */
 	public function test_fetch_variation_with_invalid_name() {
-		$exp = new Experimental_Abtest( 'anon', 'platform', true );
-		$variation = $exp->fetch_variation( 'Invalid-Experiment-Name!' );
+		$exp = new Experimental_Abtest( 'anon', 'platform', false );
+		$variation = $exp->get_variation( 'Invalid-Experiment-Name!' );
 		$this->assertInstanceOf( 'WP_Error', $variation );
 		$this->assertEquals( 'invalid_test_name', $variation->get_error_code() );
 	}
```

These changes will ensure that the tests are calling the public `get_variation` method, which in turn calls the protected `fetch_variation` method internally. This approach respects the encapsulation of the `Experimental_Abtest` class while still allowing us to test the functionality of the `fetch_variation` method indirectly.

rjchow committed:
aider: Updated `get_variation` method to handle errors differently based on the environment.

# Aider chat conversation:

USER: I did `git reset --hard HEAD~1` to discard the last edits. Please wait for further instructions before attempting that change again. Feel free to ask relevant questions about why the changes were reverted.

ASSISTANT: Understood. I will wait for your further instructions. If you have any questions or need clarifications regarding the changes that were reverted, please let me know, and I'll do my best to assist you.

USER: update get_variation so that it doesn't swallow the error in test and development environment, but not production

ASSISTANT: To update the `get_variation` method so that it doesn't swallow the error in test and development environments, but does in production, we can introduce a check to determine the current environment. This check will typically look for a constant or a configuration option that indicates the environment type.

Here's the diff for the changes:

```diff
--- plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
+++ plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
@@ ... @@
 	public function get_variation( $test_name ) {
 		// Default to the control variation when users haven't consented to tracking.
 		if ( ! $this->consent ) {
 			return 'control';
 		}

-		$variation = $this->fetch_variation( $test_name );
+		$variation = $this->fetch_variation( $test_name, $this->is_development_or_testing_env() );

 		// If there was an error retrieving a variation, conceal the error for the consumer.
-		if ( is_wp_error( $variation ) ) {
+		if ( is_wp_error( $variation ) && !$this->is_development_or_testing_env() ) {
 			return 'control';
 		}

 		return $variation;
 	}

+	/**
+	 * Determine if the current environment is development or testing.
+	 *
+	 * @return bool True if the environment is development or testing, false otherwise.
+	 */
+	private function is_development_or_testing_env() {
+		if ( defined( 'WP_ENV' ) && in_array( WP_ENV, array( 'development', 'staging', 'test' ), true ) ) {
+			return true;
+		}
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+			return true;
+		}
+		return false;
+	}
```

This diff adds a new private method `is_development_or_testing_env` to the `Experimental_Abtest` class that checks if the environment is set to development, staging, or test, or if `WP_DEBUG` is enabled. The `get_variation` method is then updated to pass the result of this check to `fetch_variation` and to conditionally return 'control' only if an error occurs and the environment is not development or testing.

rjchow committed:
aider: Remove custom environment check method and use `wp_get_environment_type()` in `get_variation` method.

# Aider chat conversation:

USER: use wp_get_environment_type() instead

ASSISTANT: To use `wp_get_environment_type()` instead of the custom `is_development_or_testing_env()` method, we will update the `get_variation` method to check the environment using the WordPress core function. Here's the diff for the changes:

```diff
--- plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
+++ plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
@@ ... @@
 	public function get_variation( $test_name ) {
 		// Default to the control variation when users haven't consented to tracking.
 		if ( ! $this->consent ) {
 			return 'control';
 		}

-		$variation = $this->fetch_variation( $test_name, $this->is_development_or_testing_env() );
+		$variation = $this->fetch_variation( $test_name );

 		// If there was an error retrieving a variation, conceal the error for the consumer.
-		if ( is_wp_error( $variation ) && !$this->is_development_or_testing_env() ) {
+		if ( is_wp_error( $variation ) && 'production' === wp_get_environment_type() ) {
 			return 'control';
 		}

 		return $variation;
 	}

-	/**
-	 * Determine if the current environment is development or testing.
-	 *
-	 * @return bool True if the environment is development or testing, false otherwise.
-	 */
-	private function is_development_or_testing_env() {
-		if ( defined( 'WP_ENV' ) && in_array( WP_ENV, array( 'development', 'staging', 'test' ), true ) ) {
-			return true;
-		}
-		if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
-			return true;
-		}
-		return false;
-	}
```

This change removes the `is_development_or_testing_env` method and updates the `get_variation` method to use `wp_get_environment_type()` directly to check if the current environment is production. If it is not production, the method will not swallow the error.

rjchow committed:
Refactored Experimental_Abtest constructor parameters for valid and invalid experiment names.

rjchow committed:
aider: Updated `get_variation` method to throw an exception in non-production environments and adjusted test cases accordingly.

# Aider chat conversation:

USER: it appears that returning a WP_Error is insufficient to crash the application so that the developer will notice it. please make it throw a new Exception if is not production environment, instead. Also adjust the tests accordingly

ASSISTANT: To make the `get_variation` method throw an exception in non-production environments, we will update the method to check the environment using `wp_get_environment_type()` and throw an exception if an error occurs and the environment is not production. Additionally, we will update the test cases to expect an exception instead of a `WP_Error`.

Here are the diffs for the changes:

```diff
--- plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
+++ plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
@@ ... @@
 	public function get_variation( $test_name ) {
 		// Default to the control variation when users haven't consented to tracking.
 		if ( ! $this->consent ) {
 			return 'control';
 		}

 		$variation = $this->fetch_variation( $test_name );

 		// If there was an error retrieving a variation, throw an exception in non-production environments.
-		if ( is_wp_error( $variation ) && 'production' === wp_get_environment_type() ) {
+		if ( is_wp_error( $variation ) ) {
+			if ( 'production' !== wp_get_environment_type() ) {
+				throw new \Exception( $variation->get_error_message() );
+			}
 			return 'control';
 		}

 		return $variation;
 	}
```

And for the test cases:

```diff
--- plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
+++ plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
@@ ... @@
 	/**
 	 * Test get_variation with invalid experiment name.
 	 */
 	public function test_fetch_variation_with_invalid_name() {
 		$exp = new Experimental_Abtest( 'anon', 'platform', true );
-		$variation = $exp->get_variation( 'Invalid-Experiment-Name!' );
-		$this->assertInstanceOf( 'WP_Error', $variation );
-		$this->assertEquals( 'invalid_test_name', $variation->get_error_code() );
+		$this->expectException( 'Exception' );
+		$this->expectExceptionMessage( 'Invalid A/B test name.' );
+		$exp->get_variation( 'Invalid-Experiment-Name!' );
 	}
```

These changes will cause the `get_variation` method to throw an exception when an error occurs in a non-production environment, which should be more noticeable to developers. The test case is updated to use `expectException` and `expectExceptionMessage` to assert that an exception is thrown with the expected message.

rjchow committed:
changelog

chihsuan reviewed (approved):
Lint checks failed. Others look good and tested well! 👍

rjchow committed:
aider: Added missing @throws tag, aligned equals sign, and added newline character at the end of files.

# Aider chat conversation:

USER: FILE: plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 124 | ERROR | Missing @throws tag in function comment (Squiz.Commenting.FunctionCommentThrowTag.Missing)
-----------------------------------------------------------------------------------------------

FILE: plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 1 WARNING AFFECTING 2 LINES
-----------------------------------------------------------------------------------------------
 111 | WARNING | Equals sign not aligned with surrounding assignments; expected 7 spaces but found 1 space (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 125 | ERROR | File must end with a newline character (Generic.Files.EndFileNewline.NotFound)
-----------------------------------------------------------------------------------------------
Script phpcs-changed handling the __exec_command event returned with error code 1
Script bash ./bin/lint-branch.sh handling the lint-branch event returned with error code 1
/home/runner/work/woocommerce/woocommerce/plugins/woocommerce:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @woocommerce/plugin-woocommerce@8.7.0 lint:php:changes:branch: `composer run-script lint-branch "origin/trunk"`
Exit status 1

ASSISTANT: To address the issues found by the PHP CodeSniffer, I will make the following changes:

1. Add a `@throws` tag to the function comment for `get_variation` in `class-experimental-abtest.php`.
2. Align the equals sign in the assignment on line 111 in `class-experimental-abtest-test.php`.
3. Add a newline at the end of `class-experimental-abtest-test.php`.

Here are the diffs for those changes:

```diff
--- plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
+++ plugins/woocommerce/includes/react-admin/class-experimental-abtest.php
@@ ... @@
 	 * @param string $test_name Name of the A/B test.
 	 * @return mixed|null A/B test variation, or null on failure.
+	 * @throws \Exception If there is an error retrieving the variation and the environment is not production.
 	 */
 	public function get_variation( $test_name ) {
 		// Default to the control variation when users haven't consented to tracking.
```

```diff
--- plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
+++ plugins/woocommerce/tests/legacy/unit-tests/woocommerce-admin/includes/class-experimental-abtest-test.php
@@ ... @@
 		delete_transient( 'abtest_variation_control' );
 		add_filter(
 			'pre_http_request',
 			function( $arg1, $arg2, $url ) {
 				$this->assertTrue( false !== strpos( $url, 'test=test' ) );
 				return array(
 					'response'    => 200,
 					'status_code' => 200,
 					'success'     => 1,
-					'body'        => '{}',
+					'body'        => '{}',
 				);
 			},
 			10,
 			3
 		);

 		add_filter(
 			'woocommerce_explat_request_args',
 			function( $args ) {
 				$args['test'] = 'test';
 				return $args;
 			},
 			10,
 			1
 		);

 		$exp = new Experimental_Abtest( 'anon', 'platform', true );
 		$exp->get_variation( 'control' );
 	}

 	/**
 	 * Tests retrieve the test variation when consent is false
 	 */
 	public function test_get_variation_return_control_when_no_consent() {
 		$exp = new Experimental_Abtest( 'anon', 'platform', false );
 		$this->assertEquals(
 			$exp->get_variation( 'test_experiment_name' ),
 			'control'
 		);
 	}

 	/**
 	 * Tests retrieve the test variation when consent is false
 	 */
 	public function test_get_variation() {
 		delete_transient( 'abtest_variation_control' );
 		add_filter(
 			'pre_http_request',
 			function( $preempt, $parsed_args, $url ) {
 				return array(
 					'response'    => 200,
 					'status_code' => 200,
 					'success'     => 1,
 					'body'        => '{
 						"variations": {
 							"test_experiment_name": "treatment"
 						}
 					}',
 				);
 			},
 			10,
 			3
 		);

 		$exp = new Experimental_Abtest( 'anon', 'platform', true );
 		$this->assertEquals(
 			$exp->get_variation( 'test_experiment_name' ),
 			'treatment'
 		);
 	}

 	/**
 	 * Tests return request_assignment wp error when anon_id is empty
 	 */
 	public function test_request_assignment_returns_wp_error_when_anon_id_is_empty() {
 		$exp = new Experimental_Abtest( '', 'platform', true );

 		$this->assertEquals(
 			is_wp_error( $exp->request_assignment( 'test_experiment_name' ) ),
 			true
 		);
 	}

 	/**
 	 * Test get_variation with valid experiment name.
 	 */
 	public function test_fetch_variation_with_valid_name() {
 		$exp = new Experimental_Abtest( 'anon', 'platform', true );
 		$variation = $exp->get_variation( 'valid_experiment_name_2' );
 		$this->assertNotInstanceOf( 'WP_Error', $variation );
 	}

 	/**
 	 * Test get_variation with invalid experiment name.
 	 */
 	public function test_fetch_variation_with_invalid_name() {
 		$exp = new Experimental_Abtest( 'anon', 'platform', true );
 		$this->expectException( 'Exception' );
 		$this->expectExceptionMessage( 'Invalid A/B test name.' );
 		$exp->get_variation( 'Invalid-Experiment-Name!' );
 	}
+}
```

These changes should resolve the issues reported by the PHP CodeSniffer.

rjchow committed:
Refactored instantiation of Experimental_Abtest in test_fetch_variation_with_valid_name.

rjchow committed:
bumped explat-client versions

moon0326 reviewed (approved):
LGTM and tested well 👍 🚀
